test/subtract_avg_cfl_test.cc

```diff
@@ -0,0 +1,142 @@
+ */
+
+/******************************************************************************
+ * @file subtract_avg_cfl_test.cc
+ *
+ * @brief Unit test for cfl subtract average function:
+ * - subtract_average_avx2
+ *
+ * @author Cidana-Edmond
+ *
+ ******************************************************************************/
+
+#include "gtest/gtest.h"
+#include "EbIntraPrediction.h"
+#include "aom_dsp_rtcd.h"
+#include "random.h"
+#include "util.h"
+
+/**
+ * @brief Unit test for cfl subtract average function:
+ * - subtract_average_avx2
+ *
+ * Test strategy:
+ * Verify this assembly code by comparing with reference c implementation.
+ * Feed the same data and check test output and reference output.
+ *
+ * Expected result:
+ * Output from assemble functions should be the same with output from c.
+ *
+ * Test coverage:
+ * Test cases:
+ * data buffer: Fill with random values
+ * TxSize: all the TxSize.
+ */
+
+#define MAKE_TEST_PARAM(function)                                            \
+    {                                                                        \
+        make_tuple(TX_4X4, &function), make_tuple(TX_4X8, &function),        \
+            make_tuple(TX_4X16, &function), make_tuple(TX_8X4, &function),   \
+            make_tuple(TX_8X8, &function), make_tuple(TX_8X16, &function),   \
+            make_tuple(TX_8X32, &function), make_tuple(TX_16X4, &function),  \
+            make_tuple(TX_16X8, &function), make_tuple(TX_16X16, &function), \
+            make_tuple(TX_16X32, &function), make_tuple(TX_32X8, &function), \
+            make_tuple(TX_32X16, &function), make_tuple(TX_32X32, &function) \
+    }
+
+namespace {
+
+using std::make_tuple;
+using svt_av1_test_tool::SVTRandom;
+
+/** Defines of test parameters */
+typedef cfl_subtract_average_fn (*get_sub_avg_fn)(TxSize tx_size);
+typedef std::tuple<TxSize, get_sub_avg_fn> sub_avg_param;
+
+class CflSubAvgTest : public ::testing::TestWithParam<sub_avg_param> {
+  public:
+    CflSubAvgTest() : tx_size_(TEST_GET_PARAM(0)) {
+        width_ = tx_size_wide[tx_size_];
+        height_ = tx_size_high[tx_size_];
+        memset(data_, 0, sizeof(data_));
+        memset(data_ref_, 0, sizeof(data_ref_));
+
+        /** get test and reference function */
+        sub_avg = TEST_GET_PARAM(1)(tx_size_);
+        sub_avg_ref = get_subtract_average_fn_c(tx_size_);
+    }
+    virtual ~CflSubAvgTest() {
+        aom_clear_system_state();
+    }
+    void prepare_data() {
+        SVTRandom rnd(15, false);
+        for (int j = 0; j < height_; j++) {
+            for (int i = 0; i < width_; i++) {
+                data_[j * CFL_BUF_LINE + i] = data_ref_[j * CFL_BUF_LINE + i] =
+                    (int16_t)rnd.random();
+            }
+        }
+    }
+    void compare_data() {
+        for (int j = 0; j < height_; j++) {
+            for (int i = 0; i < width_; i++) {
+                ASSERT_EQ(data_[j * CFL_BUF_LINE + i],
```

Comment:
If the test fail, could we dump the input data to reproduce this fail?

Reply:
it's done.